concordium-smart-contract-testing/src/impls.rs

```diff
@@ -81,33 +87,339 @@ impl ChainParameters {
+    /// let chain = ChainBuilder::new()
+    ///     .external_node_connection(Endpoint::from_static(
+    ///         "http://node.testnet.concordium.com:20000",
+    ///     ))
+    ///     .block_time_from_external()
+    ///     .build()
+    ///     .unwrap();
+    /// ```
+    pub fn block_time_from_external(mut self) -> Self {
+        self.block_time_from_external = true;
+        self
+    }
+
+    /// Build the [`Chain`] with the configured options.
+    ///
+    /// # Example
+    /// ```
+    /// # use concordium_smart_contract_testing::*;
+    ///
+    /// let chain = Chain::builder()
+    ///             // Use zero or more builder methods, for example:
+    ///             .euro_per_energy(ExchangeRate::new_unchecked(1, 50000))
+    ///             .micro_ccd_per_euro(ExchangeRate::new_unchecked(50000, 1))
+    ///             // Then build:
+    ///             .build()
+    ///             .unwrap();
+    /// ```
+    pub fn build(self) -> Result<Chain, ChainBuilderError> {
+        // Create the chain with default parameters.
+        let mut chain = Chain::new();
+
+        // Setup the external node connection if provided. This also forwards and sets
+        // the external query block.
+        if let Some(endpoint) = self.external_node_endpoint {
+            chain.setup_external_node_connection(endpoint, self.external_query_block)?;
+        }
+
+        // Check for conflicting exchange rate configurations.
+        if self.micro_ccd_per_euro.is_some() && self.micro_ccd_per_euro_from_external {
+            return Err(ChainBuilderError::ConflictingMicroCCDPerEuro);
+        }
+        if self.euro_per_energy.is_some() && self.euro_per_energy_from_external {
+            return Err(ChainBuilderError::ConflictingEuroPerEnergy);
+        }
+
+        // Set the exchange rates via an external query.
+        if self.micro_ccd_per_euro_from_external || self.euro_per_energy_from_external {
+            let exchange_rates = chain.get_exchange_rates_via_external_node()?;
+            if self.micro_ccd_per_euro_from_external {
+                chain.parameters.micro_ccd_per_euro = exchange_rates.micro_ccd_per_euro;
+            }
+            if self.euro_per_energy_from_external {
+                chain.parameters.euro_per_energy = exchange_rates.euro_per_energy;
+            }
+        }
+
+        // Set the exchange rates directly.
+        if let Some(micro_ccd_per_euro) = self.micro_ccd_per_euro {
+            chain.parameters.micro_ccd_per_euro = micro_ccd_per_euro;
+        }
+        if let Some(euro_per_energy) = self.euro_per_energy {
+            chain.parameters.euro_per_energy = euro_per_energy;
+        }
+
+        // Check the exchange rates and return early if they are invalid.
+        check_exchange_rates(
+            chain.parameters.euro_per_energy,
+            chain.parameters.micro_ccd_per_euro,
+        )?;
+
+        match (self.block_time, self.block_time_from_external) {
```

Comment:
Couldn't we prevent this already in the builder while calling these methods? And so we would not have to fail here in `build`?

Reply:
We could, but then we would have to return a `Result` in some of the configuration methods (but not all), which I don't think is ideal. As far as I could tell, the typical way is to only have the build method return a `Result`.
Or do you have another way of doing it in mind?

Reply:
No, that would be the result. I thought using `?` would make it pretty painless.

But I'm ok the way it is if you think it's better. I guess in practice the error will be in the "same place", when they run.

Reply:
Ok, then let's keep it as is.